### DISCUSSION TITLE
Questions of program "InstrumentStarCamera2RollPitchYaw"

### DISCUSSION THREAD

**Discussion by FFFFFG-IGG** [groops-devs/groops#214] [Q&A]
Hello developers,

  When I use program "InstrumentStarCamera2RollPitchYaw"  to get the yaw angle from obx file, I find yaw of some BDS satellites does not have correct pattern (extract attitude file with program 'GNSSOrbex2StarCamera', then get yaw angle by 'yaw*180/3.14'):
![Y@5}2``YNE(57%)T5U2CRX6](https://github.com/groops-devs/groops/assets/59271400/3b3e0e67-fcc5-44b9-9759-7f3cda10af12)
the correct attitude should be:
![1B9Q 9A$7~R0O1AKZQ}_06T](https://github.com/groops-devs/groops/assets/59271400/a15b5222-0a3c-455e-be77-1035136b87fa)
So I speculate there may be something wrong with this program and it can not get the correct yaw angle.
Data and xml are listed in the attachment.
[Uploading obx.zip…]()

regards,
Yongzhao Fan

**Comment by Waley04** [comment]
@FFFFFG-IGG 

I tried PRN46 for BDS on DOY 268, 2023. Here is what I obtained:

**== prnC46_doy268 ==**
[prnC46_doy268.txt](https://github.com/groops-devs/groops/files/12743263/prnC46_doy268.txt)

**== Fig_1 ==**
![1](https://github.com/groops-devs/groops/assets/101952777/3487cd69-d844-410d-98be-80e23ef5eb9f)

**== Fig_2 ==**
![2](https://github.com/groops-devs/groops/assets/101952777/b22010ef-99ff-46e4-a332-830347ad34de)

**== Fig_3 ==**
![3](https://github.com/groops-devs/groops/assets/101952777/18a15ca5-4e8e-472b-8999-f9eeedb54c60)

While we are waiting for support from the developers, could you try to see if you get similar results?

Regards,

Waley.

**Reply by FFFFFG-IGG** [reply]
Hi, Waley

I've gotten similar results when comparing thoes of yours, the roll, pitch, yaw angles of what I got is:

![C46_yawAngle](https://github.com/groops-devs/groops/assets/59271400/45e62c39-f4ef-436b-857a-543fbcd49a2d)

And the result referenced to yours is:

![C46_yawAngle_compare](https://github.com/groops-devs/groops/assets/59271400/90ffd8a3-ed87-4b97-b959-8f35af73c2b4)

So it may be reasonable to speculate that there are some problems in the program "InstrumentStarCamera2RollPitchYaw". Some suggestions or corrections from developers may be of help.

regards,

Yongzhao Fan

**Reply by zhedumi** [reply]
Hi,

I just realized that the yaw angle for that satellite seems very weird. C46 if i remember correct uses yaw steering and should never show this kind of constant yaw angle? It should always show some rotations in +/- depending on the beta angle to the sun if i got that right in mind.


best regards,

Patrick

**Comment by tmayerguerr** [comment]
Hi, 

The angles are computed in GROOPS from the rotary matrix R_ik as
``` 
 roll  = -atan2( R[2][1], R[2][2])
 pitch = -atan2(-R[2][0], sqrt(R[2][1]^2+R[2][2]^2))
 yaw   = -atan2( R[1][0], R[0][0])
```
I don't know if these are the right equations or if they are stable for large angles.

regards
Torsten

**Comment by zhedumi** [comment]
Hi,

or some reason the .zip file does not allow me to download. I will have a look into it as soon as I have the same data. Could you upload the OBX file alone so I may test it with the same data?

regards,

Patrick

**Reply by FFFFFG-IGG** [reply]
Hello Patrick,

[WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz](https://github.com/groops-devs/groops/files/12817145/WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz)
Here is OBX file downloaded from whu PPP-AR work-group, hope you can find some way to fix this issue!

regards,
Yongzhao Fan

**Comment by zhedumi** [comment]
So I run some tests with this file and I cannot really replicate what you guys computed.
> [WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz](https://github.com/groops-devs/groops/files/12817145/WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz) 

So I converted this file using [GnssOrbex2StarCamera](https://groops-devs.github.io/groops/html/GnssOrbex2StarCamera.html) followed by [InstrumentStarCamera2RollPitchYaw](https://groops-devs.github.io/groops/html/InstrumentStarCamera2RollPitchYaw.html). And to have another comparison I simulated the orbits with broadcast orbits and computed the attitude with [SimulateStarCameraGnss](https://groops-devs.github.io/groops/html/SimulateStarCameraGnss.html) followed by [InstrumentStarCamera2RollPitchYaw](https://groops-devs.github.io/groops/html/InstrumentStarCamera2RollPitchYaw.html) again.

This is the reuslt with simulating the starcamera
![rollpitchyaw C46](https://github.com/groops-devs/groops/assets/77068736/cf5d4c30-d801-4115-a491-bab24e758a85)


This is the result using the orbex file
![rollWithEarth C46](https://github.com/groops-devs/groops/assets/77068736/fc452181-01d9-4c25-977d-3e515ca3caa2)

What I did find out is that there is some error for Satellites in Geostationary orbits.
So I converted one time the obx file using 
[GnssOrbex2StarCamera](https://groops-devs.github.io/groops/html/GnssOrbex2StarCamera.html)  with earth rotation and one time without. Followed by [GnssOrbex2StarCamera](https://groops-devs.github.io/groops/html/GnssOrbex2StarCamera.html). TO convert them back to the original. For the BeiDou stationary satellites the attitude quantiones are faulty as well as for C11 and C12 for some reason?

This is how the difference should look like between transforming to attitude and back to OBX file:
![quaternionsDiff C07](https://github.com/groops-devs/groops/assets/77068736/3b337b86-02a4-469a-b585-03cf2b6d2528)

This is how it looks for C01-C05 and C11-C12: 
![quaternionsDiff C02](https://github.com/groops-devs/groops/assets/77068736/72192417-ba45-4a92-86e8-279bf51536ce)
![quaternionsDiff C11](https://github.com/groops-devs/groops/assets/77068736/d47b0645-e667-4abe-89d9-fc7b05023971)

I am currently not sure why the inclusion of the earth rotation would screw up the attitudes for those satellites and I am still investigating.  
On another note could one of you show me the script/earth rotations you used for the attitude conversion so I can completely replicate what you did?

Best regards,
Patrick

**Reply by Waley04** [reply]
Hi @zhedumi,

I was wondering if you had resolved this issue.

Any updates?

**Reply by zhedumi** [reply]
Hi Waley,

No, I was a bit occupied by other things sadly and actually forgot about this one. I apologize and thank you for reminding me! I could not replicate your attitudes and I recogized that your attitudes behave a bit weird if I remember correctly the attitude behaviour of BDS? (check my comment here: https://github.com/groops-devs/groops/discussions/214#discussioncomment-7252806).
The other issue why the conversion from OBX to attitude results in weird behaviour using the earth rotation is something that I had written up but completely distracted by other things. 

I will check tomorrow thoroughly again and will give an update tomorrow if I found some other error in the conversion or rotation with the earth rotation! Sorry for the delay. It might even be already fixed since there was work done for the rotation in recent months.


Best regards,

Patrick

**Comment by zhedumi** [comment]
Hi,

So I did a more thorough investigation and can say that the earth rotation is not an issue. It seems to be that the data depending from which analysis center processed, lead sto this behaviour. 
I presume they put a certain number of significant digits into it. 
Basically what I did which is described here https://github.com/groops-devs/groops/discussions/214#discussioncomment-7252614 is to just read them in, convert them to attitudes and then convert them back and check if there are any errors. There are some weird patterns for different data sources but it seems more on numerical issue along with precision of the data given?
[COD0MGXFIN_20230510000_01D_30S_ATT.OBX.gz](https://github.com/groops-devs/groops/files/13843151/COD0MGXFIN_20230510000_01D_30S_ATT.OBX.gz) for example results in
![quaternionsDiff C11](https://github.com/groops-devs/groops/assets/77068736/681cc378-2dc8-4b12-bd61-c2eda1af8e78)

[WUM0MGXFIN_20230510000_01D_30S_ATT.OBX.gz](https://github.com/groops-devs/groops/files/13843168/WUM0MGXFIN_20230510000_01D_30S_ATT.OBX.gz) leads to
![quaternionsDiff C11](https://github.com/groops-devs/groops/assets/77068736/c29649b6-f374-4e71-a24c-3d47f7decb10)

[WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz](https://github.com/groops-devs/groops/files/13843171/WUM0MGXRAP_20230510000_01D_30S_ATT.OBX.gz) leads to
![quaternionsDiff C11](https://github.com/groops-devs/groops/assets/77068736/235f3a27-d252-429d-8063-25e4ab50c848)

Since COD and WUM_final show the same behaviour I would presume that the earthrotation issue is not an issue but more so that the data sources have different precision and that might lead to nummerical differences when doing the rotations etc. Or at least I cannot find a different reason as to why COD-WUM finals show the same behaviour and WUM rapid has some jumps in it while both FIN and RAP form WUM have the same sampling rate etc. Very weird but for now I presume its just nummerics and not that important.
So for those three I compared the attitudes with the COD_FIN products to see the differences. I used C46 so it is consistent with all the other attitude plots in this discussion.
EDIT: plotted the wrong thing here againt he right ones
![diffAttitudesRoll](https://github.com/groops-devs/groops/assets/77068736/bae898a2-560c-4014-8390-71f56bb3e691)
![diffAttitudesPitch](https://github.com/groops-devs/groops/assets/77068736/45e8ea24-8509-4df0-a73d-a34c5fd0bac0)
![diffAttitudesYaw](https://github.com/groops-devs/groops/assets/77068736/058773a7-6bc3-42cf-81dc-e923a242d3a0)

There are some discrepancy but I would say they are so small that they can be disregarded.
Regarding the behaviour of yaw: I get the same behaviour as linked here https://github.com/groops-devs/groops/discussions/214#discussioncomment-7252614 for all three. 
But here are the attitudes from COD fin just to show again:
![rollWithEarth C46](https://github.com/groops-devs/groops/assets/77068736/da2682ee-c2c5-4d0f-bc62-eb2a65bc59c0).

So I cannot replicate the issues you guys have with attitude at the moment. So if possible I would need the script that you use to compute the attitudes. As I noted within https://github.com/groops-devs/groops/discussions/214#discussioncomment-7252806 the yaw angle looks weird for you guys since it definitly should show more rotations.

Cheers,
Patrick
